### ISSUE TITLE
[Unix] VideoMode when rotated incorrect

### ISSUE THREAD

**Issue by h0tw1r3** [LaurentGomila/SFML#771]
VideoMode does not take into consideration rotation.

For example, I am running X at 1600x1200, but xrandr rotated 90 deg, so the effective resolution is 1200x1600.

Impl just needs updated to respect current and possible rotation resolutions (x/y flipped on all if supported by xrandr).

**Comment by TankOs** [comment]
At first, you are talking about fullscreen, right?

SFML uses XCB to find a suiting video mode for the user's requested mode. It then uses that mode's rotation setting when applying it. It might be possible that there are several modes with different rotation settings, but same width and height properties.

I will check this, thanks for reporting.

**Comment by h0tw1r3** [comment]
#include <SFML/Graphics.hpp>
    #include <iostream>

    int main() {
        sf::VideoMode mode = sf::VideoMode::getDesktopMode();
        std::cout << mode.width << " x " << mode.height << std::endl;
    }

Expected output is: 1200 x 1600
Actual output: 1600 x 1200

**Comment by h0tw1r3** [comment]
Additionally, getFullScreenModes() doesn't include the rotated mode.  Not sure if it should, but even so - if I try to set the mode to 1200 x 1600 it doesn't let me because it's not considered valid.

        std::vector<sf::VideoMode> modes = sf::VideoMode::getFullscreenModes();

        for (unsigned i = 0; i < modes.size(); i++) {
                if (!modes[i].isValid())
                        std::cout << "Invalid:";
                std::cout << i << " - " << modes[i].width << " x " << modes[i].height << " x " << modes[i].bitsPerPixel << std::endl;
        }

**Comment by TankOs** [comment]
I did some research. The rotation setting is not part of the video mode at all, it's a setting for the X display. The latest master uses XCB for changing the resolution, but apparently that's broken at the moment (a fix is being prepared). However, with XCB we use the same rotation config that's being reported by XCB, so it theoretically should work with your system as well.

I will report back when the fullscreen issue is fixed. You can then build the latest SFML yourself and see if it fixes your problem. Earliest release that contains the fix will be 2.3, though.

**Comment by h0tw1r3** [comment]
I made a slight change to get up and going.  I'm testing off of the master branch.

Obviously this will break fullscreen itself - even if it was working.

Based on how things are currently structured, should the rotation tracking be tucked away in WindowX11Impl? If so, then the VideoModes should be populated as such (all reversed) for application use.

    diff --git a/src/SFML/Window/Unix/VideoModeImpl.cpp b/src/SFML/Window/Unix/VideoModeImpl.cpp
    index 2a3489f..8cf79b5 100644
    --- a/src/SFML/Window/Unix/VideoModeImpl.cpp
    +++ b/src/SFML/Window/Unix/VideoModeImpl.cpp
    @@ -144,7 +144,10 @@ VideoMode VideoModeImpl::getDesktopMode()
                     int nbSizes = xcb_randr_get_screen_info_sizes_length(config);
                     xcb_randr_screen_size_t* sizes= xcb_randr_get_screen_info_sizes(config);
                     if (sizes && (nbSizes > 0))
    -                    desktopMode = VideoMode(sizes[currentMode].width, sizes[currentMode].height, screen->root_depth);
    +                    if (config->rotation == XCB_RANDR_ROTATION_ROTATE_90 || config->rotation == XCB_RANDR_ROTATION_ROTATE_270)
    +                        desktopMode = VideoMode(sizes[currentMode].height, sizes[currentMode].width, screen->root_depth);
    +                    else
    +                        desktopMode = VideoMode(sizes[currentMode].width, sizes[currentMode].height, screen->root_depth);
                 }
                 else
                 {

**Comment by h0tw1r3** [comment]
Made a few updates to the code in my previous comment.  If it looks worthwile, I will submit a pull request.

**Comment by LaurentGomila** [comment]
```std::swap``` is your friend ;)

**Comment by TankOs** [comment]
I see, however I wonder by xcb_set_screen_config() asks for a rotation as well. Like said, at first I have to fix the outstanding issues before we can continue on this. Your patch might help, though, so thanks in advance.

**Comment by h0tw1r3** [comment]
@LaurentGomila :) yep, ++ is not in my wheelhouse yet. Thanks for pointing that out.

**Comment by TankOs** [comment]
@h0tw1r3 Please see #780, it should fix your issue.

**Comment by h0tw1r3** [comment]
@TankOs Tested, but does not fix the problem here.  My monitor is rotated by defailt (Vertical).  I am typing this message from an X resolution of 1920x1200 - rotated left (270 deg.).  Effective resolution is 1200x1920.

xrandr rotate your screen, then run the test program above. It will return the non-rotated resolution.  Which would be fine, but rotation is not exposed in the api - so it can't be accounted for by the application.

The only fix I can see is to either expose the current screen rotation, or hide it and account for it (as my patch above does).

**Comment by h0tw1r3** [comment]
Fullscreen is not what I'm going for.  Just need to get/set the current (effective?) desktop video mode / resolution.

**Comment by binary1248** [comment]
@h0tw1r3: In #780 Tank rewrote how fullscreen switches are performed, but didn't touch the mode enumeration code or the code to retrieve the current (possibly rotated) mode. Perhaps you could consider submitting a PR containing your changes to VideoModeImpl.cpp, i.e. the first 2/3 of your diff?

**Comment by binary1248** [comment]
Fixed in #825.

**Comment by binary1248** [comment]
Merged in b9cc6f5c0de4587c70ae9447a1d7a501cefaded1
